### PR TITLE
AppVeyor updates & evict the configure cache if `configure` changed 

### PR DESCRIPTION
**Pull request by MisterDA**
This PR sports updates and fixes for AppVeyor. Amongst the noticeable changes:

- move to Visual Studio 2022 by default;
- AppVeyor will only trigger on branch pushes which names match `4\.*`, `5\.*`, `trunk`, and all pull requests pushes, just as GitHub Actions;
- evict the configure cache if configure has changed: the script responsible for a cache entry might have been modified. Detect whether configure has changed by computing a checksum and retrieving the system version, and evict the cache if needed.

### CODE REVIEW

File: tools/ci/appveyor/appveyor_build.sh [hunk 1]
```diff
@@ -83,11 +81,24 @@ function set_configuration {
...
 
     mkdir -p "$CACHE_DIRECTORY"
-    ./configure --cache-file="$CACHE_DIRECTORY/config.cache-$1" \
-                $dep $build $man $host --prefix="$2" --enable-ocamltest || ( \
-      rm -f "$CACHE_DIRECTORY/config.cache-$1" ; \
-      ./configure --cache-file="$CACHE_DIRECTORY/config.cache-$1" \
-                  $dep $build $man $host --prefix="$2" --enable-ocamltest )
+
+    local CACHE_KEY CACHE_FILE_PREFIX CACHE_FILE
+    CACHE_KEY=$(cat configure <(uname) | shasum | cut -c 1-7)
```
Comment (nojb) [inline tools/ci/appveyor/appveyor_build.sh]: Simpler:
```suggestion
    CACHE_KEY=$({ cat configure; uname; } | shasum | cut -c 1-7)
```

**Comment by shindere** [comment]
I'm far from knowing these things but from a distance this PR looks very
good to me.

I just have two questions:

1. In the last commit about the cache there is this comment:

# Remove old configure cache if the configure script has changed

Could the comemntperhaps be made more precise by including that an OS
change also invalidates the cache?

2. In the first commit, if I read the diff correctly the `exit /b 1` was
done after the files were downloadid befre the PR and with the PR it's
done before the file get dowloaded and I don't understandthis change.

File: tools/ci/appveyor/appveyor_build.sh (hunk 1 above)
Comment (MisterDA) [reply tools/ci/appveyor/appveyor_build.sh]: Applied, thanks!

**Comment by MisterDA** [comment]
> In the last commit about the cache there is this comment: # Remove old configure cache if the configure script has changed Could the comment perhaps be made more precise by including that an OS change also invalidates the cache?

Done.

> In the first commit, if I read the diff correctly the `exit /b 1` was done after the files were downloaded before the PR and with the PR it's done before the file get downloaded and I don't understand this change.

We've clarified this offline, and it was a problem stemming from the long lines, but the code is fine as-is.

I've applied all the suggestions.

**Comment by shindere** [comment]
ThePR looks good to me and it's changing only our CI so the risk of any
problem lookslow to me.

Thus approving and merging.

**Review by shindere — APPROVED** [review]

File: tools/ci/appveyor/appveyor_build.sh [hunk 2]
```diff
@@ -26,6 +26,8 @@ else
   MAKE=make
 fi
 
+git config --global --add safe.directory '*'
```
Comment (nojb) [inline tools/ci/appveyor/appveyor_build.sh:29]: Why was this not necessary before?

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2023/10/30 10:28 -0700):
> @nojb commented on this pull request.
>
>
>
> > @@ -26,6 +26,8 @@ else
>    MAKE=make
>  fi
>
> +git config --global --add safe.directory '*'
>
> Why was this not necessary before?

Oops. @nojb if you thinkthe merge was premature feel free to either tell
me or revert it yourself. Sorry.

**Comment by nojb** [comment]
> Oops. @nojb if you thinkthe merge was premature feel free to either tell me or revert it yourself. Sorry.

No worries!

File: tools/ci/appveyor/appveyor_build.sh (hunk 2 above)
Comment (MisterDA) [reply tools/ci/appveyor/appveyor_build.sh:29]: Better late than never: this was due to a Git CVE.
https://github.blog/2022-04-12-git-security-vulnerability-announced/#cve-2022-24765
To mitigate, Git started issuing warnings on Windows (in some cases) if it encounters a new Git repository. Affected users have to register globally all paths leading to Git repositories.
In our CI we don't care, so we can allow everything. It prevents git from exiting with an error status.

**Pull request merged**